pkg/coscheduling/coscheduling.go

```diff
@@ -95,294 +83,364 @@ func New(obj runtime.Object, handle framework.FrameworkHandle) (framework.Plugin
 	return nil
 }
 
 // Unreserve rejects all other Pods in the PodGroup when one of the pods in the group times out.
 func (cs *Coscheduling) Unreserve(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) {
-	pgInfo, _ := cs.getOrCreatePodGroupInfo(pod, time.Now())
-	if len(pgInfo.key) == 0 {
+	_, pgs, _ := cs.pgMgr.GetPodGroup(pod)
+	if pgs == nil {
 		return
 	}
-	podGroupName := pgInfo.name
-	cs.frameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
-		if waitingPod.GetPod().Namespace == pod.Namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
-			klog.V(3).Infof("Unreserve rejects the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
-			waitingPod.Reject(cs.Name())
+	pgs.AssumedPods.Delete(core.GetPodName(pod))
+}
+
+// PostBind is called after a pod is successfully bound. These plugins are used update PodGroup when pod is bound.
+func (cs *Coscheduling) PostBind(ctx context.Context, _ *framework.CycleState, pod *v1.Pod, nodeName string) {
+	klog.V(5).Infof("PostBind pod: %v/%v", pod.Namespace, pod.Name)
+	cs.pgMgr.PostBind(ctx, pod, nodeName)
+}
+
+// VerifyLocalCache verify pgstatus cache
+func (cs *Coscheduling) VerifyLocalCache(toSchedulePGName string) {
+	pgs := cs.pgMgr.PodGroupStatusCache().Get(toSchedulePGName)
+	// update pods cache
+	cs.frameworkHandler.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+		klog.V(5).Info("Start IterateOverWaitingPods")
+		pod := waitingPod.GetPod()
+		pgName := util.VerifyPodLabelSatisfied(pod)
+		klog.V(5).Infof("Start walking through pod %v/%v uid: %v", pod.Namespace, pod.Name, pod.UID)
+		if pgName == toSchedulePGName {
+			klog.V(5).Infof("Wanted cache pod %v/%v", pod.Namespace, pod.Name)
+			_, found := pgs.AssumedPods.Get(core.GetPodName(pod))
+			if !found {
+				return
+			}
+			klog.V(6).Infof("Current pod group state %+v", pgs.AssumedPods.Items())
+			cs.pgMgr.PodGroupStatusCache().Set(util.GetPodGroupFullName(pgs.PodGroup), pgs)
 		}
 	})
 }
 
-// GetPodGroupLabels checks if the pod belongs to a PodGroup. If so, it will return the
-// podGroupName, minAvailable of the PodGroup. If not, it will return "" and 0.
-func GetPodGroupLabels(pod *v1.Pod) (string, int, error) {
-	podGroupName, exist := pod.Labels[PodGroupName]
-	if !exist || len(podGroupName) == 0 {
-		return "", 0, nil
+// StartBatchSchedule receives the group name of pod to start scheduling
+func (cs *Coscheduling) StartBatchSchedule(fullName string) {
+	pgs := cs.pgMgr.PodGroupStatusCache().Get(fullName)
+	if pgs.PodGroup.Status.Phase != v1alpha1.PodGroupPreScheduling && pgs.PodGroup.Status.Phase != v1alpha1.
+		PodGroupScheduling {
+		return
+	}
+
+	// record time to avoid abnormal exit when bind
+	if pgs.PodGroup.Status.Scheduled >= pgs.PodGroup.Spec.MinMember {
+		pg := pgs.PodGroup
+		pgCopy := pgs.PodGroup.DeepCopy()
+		pgCopy.Status.ScheduleStartTime = metav1.Now()
+
+		patch, err := util.CreateMergePatch(pg, pgCopy)
```

Comment:
In CreateMergePatch(), prefer to use strategicpatch.CreateTwoWayMergePatch() instead of importing an external library.